File: 프로그래머스/KaKao/표 병합.py
def solution(commands):
    answer = []
    table = [["EMPTY"] * 51 for _ in range(51)]
    merged = [[(i, j) for j in range(51)] for i in range(51)]

    for command in commands:
        cmd = command.split(" ")

        if cmd[0] == "UPDATE":
            if len(cmd) == 4:
                r, c = merged[int(cmd[1])][int(cmd[2])]
                table[r][c] = cmd[3]

            elif len(cmd) == 3:
                for i in range(1, 51):
                    for j in range(1, 51):
                        if table[i][j] == cmd[1]:
                            table[i][j] = cmd[2]

        elif cmd[0] == "MERGE":
            r1, c1 = merged[int(cmd[1])][int(cmd[2])]
            r2, c2 = merged[int(cmd[3])][int(cmd[4])]

            if r1 == r2 and c1 == c2:
                continue

            if table[r1][c1] == "EMPTY" and table[r2][c2] != "EMPTY":
                table[r1][c1] = table[r2][c2]

            for i in range(1, 51):
                for j in range(1, 51):
                    if merged[i][j] == (r2, c2):
                        merged[i][j] = (r1, c1)

        elif cmd[0] == "UNMERGE":
            r, c = merged[int(cmd[1])][int(cmd[2])]
            preValue = table[r][c]

            for i in range(1, 51):
                for j in range(1, 51):
                    if merged[i][j] == (r, c):
                        merged[i][j] = (i, j)
                        table[i][j] = "EMPTY"
            table[int(cmd[1])][int(cmd[2])] = preValue

        elif cmd[0] == "PRINT":
            r, c = merged[int(cmd[1])][int(cmd[2])]
            answer.append(table[r][c])

    return answer

File: 프로그래머스/KaKao/표병합.swift
// 2023 KAKAO BLIND RECRUITMENT 표 병합
import Foundation

func solution(_ commands:[String]) -> [String] {
    var answer:[String] = []
    var table:[[String]] = Array(repeating: Array(repeating: "EMPTY", count: 50), count: 50)
    var merged:[[(Int,Int)]] = []
    
    for i in 0..<50 {
        var line: [(Int, Int)] = []
        for j in 0..<50 {
            line.append((i, j))
        }
        merged.append(line)
    }
    
    for cmd in commands {
        let splitedCmd = cmd.components(separatedBy: " ")
        //print(splitedCmd)
        
        if splitedCmd[0] == "UPDATE" {
            if splitedCmd.count == 4 {
                let m = merged[Int(splitedCmd[1])!-1][Int(splitedCmd[2])!-1]
                table[m.0][m.1] = splitedCmd[3]
            } else {
                for i in 0..<50 {
                    for j in 0..<50 {
                        if table[i][j] == splitedCmd[1] {
                            table[i][j] = splitedCmd[2]
                        }
                    }
                }
            }
        } else if splitedCmd[0] == "MERGE" {
            let n1 = merged[Int(splitedCmd[1])!-1][Int(splitedCmd[2])!-1]
            let n2 = merged[Int(splitedCmd[3])!-1][Int(splitedCmd[4])!-1]
            
            if n1 == n2 {
                continue
            }
            
            if table[n1.0][n1.1] == "EMPTY" {
                table[n1.0][n1.1] = table[n2.0][n2.1]
            }
            
            for i in 0..<50 {
                for j in 0..<50 {
                    if merged[i][j] == n2 {
                        merged[i][j] = n1
                    }
                }
            }
            
        } else if splitedCmd[0] == "UNMERGE" {
            let m = merged[Int(splitedCmd[1])!-1][Int(splitedCmd[2])!-1]
            let preVal = table[m.0][m.1]
            
            for i in 0..<50 {
                for j in 0..<50 {
                    if merged[i][j] == m {
                        merged[i][j] = (i, j)
                        table[i][j] = "EMPTY"
                    }
                }
            }
            table[Int(splitedCmd[1])!-1][Int(splitedCmd[2])!-1] = preVal
            
        } else if splitedCmd[0] == "PRINT" {
            let m = merged[Int(splitedCmd[1])!-1][Int(splitedCmd[2])!-1]
            answer.append(table[m.0][m.1])
        }
    }
    
    return answer
}